[docs] Add advanced page for the system (#23596)
mui-org/material-ui: docs/src/pages/system/advanced/advanced.md

# Advanced

<p class="description">Here you can find examples of how you can use the system in your custom components.</p>

## Adding the `sx` prop to your custom components

> The `unstable_styleFunctionSx` utility adds the support for the `sx` to your own components. You may use the `Box` components that comes from `@material-ui/core` as a root of your component tree instead, but if you would not like to depend on the core and have a smaller bundle size, this utility would give you the same capabilities.

Reviewer: ```suggestion
The `unstable_styleFunctionSx` utility adds the support for the `sx` to your own components. Normally you would use the `Box` components from `@material-ui/core` as the root of your component tree. If you would like to use it independently from Material-UI, this utility will give you the same capabilities, while having a smaller bundle size.
```

> The `unstable_styleFunctionSx` utility adds the support for the `sx` to your own components. Normally you would use the `Box` components from `@material-ui/core` as the root of your component tree. If you would like to use it independently from Material-UI, this utility will give you the same capabilities, while having a smaller bundle size.

Reviewer: ```suggestion
The `unstable_styleFunctionSx` utility adds the support for the `sx` to your own components. Normally you would use the `Box` components from `@material-ui/core` at the root of your component tree. If you would like to use the system independently from Material-UI, this utility will give you the same capabilities, while having a smaller bundle size.
```

{{"demo": "pages/system/advanced/StyleFunctionSxDemo.js"}}

## Using standalone system utilities

> If you wouldn't like to use the `sx` prop on your custom components, but only some bits from the system, you can use and combine the different style functions available as part of the system. You are likely to use this approach if you neeed smaller bundle size and better performance for the price of using a small set of what the `sx` supports.

Reviewer: ```suggestion
If you only need some elements of the system in your custom components, you can directly use and combine the different style functions available, and access them as component props. You might use this approach if you need smaller bundle size and better performance than using Box, for the price of using a subset of what the `sx` supports, and a different API.
```

# Advanced

<p class="description">Here you can find examples of how you can use the system on your custom components.</p>

## Adding the `sx` prop on your custom components

The `unstable_styleFunctionSx` utility adds the support for the `sx` in your own components.

{{"demo": "pages/system/advanced/StyleFunctionSxDemo.js"}}

> ### Built-in `sx` from the core

Author: The advantage is that you can add it to more complex custom components, and it would behave is it does on the core MUI components. 

I had this initially as first secrtion, as it seemd like the easiest way to add it, but @oliviertassinari had a valid point that this is a page for the system and the first thing we are advertising is something from the core, that’s why we moved it as second section. I am ok with any order, but I think I should probably rephrase to explain why it is valuable (if it is not obvious to you guys, than this is a bad docs from my side :/)

> ### Built-in `sx` from the core

Reviewer: > you can add it to more complex custom components

I think the point is that in this instance (already using core) you could just use Box as your root component.

> @oliviertassinari had a valid point that this is a page for the system and the first thing we are advertising is something from the core

Notwithstanding the above, it seems to be about how to use the styles system without using Box.

> ### Built-in `sx` from the core

Reviewer 2: Let's rewind a bit, there are multiple layers, by level of abstractions:

1. the system functions
2. the styled engine
3. the styled() implementation
4. the theme
5. the Box
6. the usage of `sx`

This echo to a discussion we had about moving Box into the system.

> ### Built-in `sx` from the core

Reviewer: > the styled engine
"style engine". The sooner we rename that the better.

That aside, having rewound (and played forward) what was your point?

> ### Built-in `sx` from the core

Reviewer 3: > @eps1lon - don't want to put words in your mouth here...

That is my point. I'm just completely lost why I would use this functionality. The issues described are not voiced by any users (how could they if they didn't have time to use `sx`) and the API itself hasn't had any time to incubate. But we're already pushing everything we can to production level without having documentation. The pages added just showcase usage which is trivial.

In my opinion our internal styled should not be published at this point. We didn't have any time to battle-test the API.

But I think I've already said this enough so there's no point repeating anymore. Just said it because I was pinged explicitly.

> ### Built-in `sx` from the core

Author: Should we then hold off with these changes until we have an actual requirement for something like this?

> ### Built-in `sx` from the core

Author: I am either way describing experimental and unstable API so maybe we should just wait and see the feedback on the prop first...

> ### Built-in `sx` from the core

Reviewer 2: There is one aspect I would like to raise: listening to users has limits too. No customer [ever asked](https://www.sec.gov/Archives/edgar/data/1018724/000119312517120198/d373368dex991.htm) Amazon to build Amazon prime, they have done it anyway. I think that what's fundamentally important is either or not can we envision this abstraction to be used or not? In our case, more by front-end developers or designers (we have a large audience of full-stack developers, so it doesn't help).

> ### Built-in `sx` from the core

Reviewer: As a compromise, how about removing this section ("Built-in `sx` from the core") for now? Sure, it's another way of achieving the same end result, but it doesn't seem to add any value over just using Box. Presenting multiple ways to achieve the same thing just leads to confusion.

In the other two sections, we emphasise the benefits of each approach. (Use without core / less overhead, respectively.)

> ### Built-in `sx` from the core

Author: Removed that section and added reasons of why someone may use these utilities. I hope is much clear now :)

> The easiest way to support the `sx` prop on your custom components is to use the `experimentalStyled()` utility that comes from `@material-ui/core/styles` for creating your custom components. Here is an example of how it can be use:

Reviewer 2: ```suggestion
The easiest way to support the `sx` prop on your custom components is to use the `experimentalStyled()` utility that comes from `@material-ui/core/styles` for creating your custom components. Here is an example of how it can be used:
```

> {{"demo": "pages/system/advanced/ExperimentalStyledDemo.js"}}

Reviewer 2: I think that it's short enough to display the whole content

```suggestion
{{"demo": "pages/system/advanced/ExperimentalStyledDemo.js", "defaultCodeOpen": true}}
```

> If you are already using the Material-UI's `styled()` wrapper, the easiest way to support the `sx` prop on your custom components is to use the `experimentalStyled()` utility that comes from `@material-ui/core/styles` for creating your custom components. Here is an example of how it can be used:

Reviewer: ```suggestion
If you're already using Material-UI's `styled()` wrapper, the easiest way to support the `sx` prop in your custom components is to use the `experimentalStyled()` utility that comes from `@material-ui/core/styles`. Here is an example of how it can be used:
```

> The `experimentalStyled()` offers other functionalities, like access to default theme, as well as access to the theme overrides and variants. If you would not like to use these utility, you may use the lighter version of utility for adding the `sx` utility on your custom components - the `unstable_styleFunctionSx` utility.

Reviewer 2: ```suggestion
The `experimentalStyled()` offers other functionalities, like access to the default theme, as well as access to the theme overrides and variants. If you would not like to use these utilities, you may use the lighter version of the utility for adding the `sx` utility on your custom components - the `unstable_styleFunctionSx` utility.
```

> ## Adding the `sx` prop on your custom components
> 
> The easiest way to support the `sx` prop on your custom components is to use the `experimentalStyled()` utility that comes from `@material-ui/core/styles` for creating your custom components. Here is an example of how it can be used:
> 
> {{"demo": "pages/system/advanced/ExperimentalStyledDemo.js", "defaultCodeOpen": true}}
> 
> ## Use the `unstable_styleFuntionSx` utility
> 
> The `experimentalStyled()` offers other functionalities, like access to the default theme, as well as access to the theme overrides and variants. If you would not like to use these utilities, you may use the lighter version of the utility for adding the `sx` utility on your custom components - the `unstable_styleFunctionSx` utility.
> 
> {{"demo": "pages/system/advanced/StyleFunctionSxDemo.js"}}

Reviewer 2: We might want to focus on `@material-ui/system`. I would challenge any import from `@material-ui/core`.

```suggestion
## Adding the `sx` prop on your custom components

The `unstable_styleFunctionSx` utility adds the support for the `sx` in your own components.

{{"demo": "pages/system/advanced/StyleFunctionSxDemo.js"}}

### Built-in `sx` from the core

If you are already using the Material-UI's `styled()` wrapper, the easiest way to support the `sx` prop on your custom components is to use the `experimentalStyled()` utility that comes from `@material-ui/core/styles` for creating your custom components. Here is an example of how it can be used:

{{"demo": "pages/system/advanced/ExperimentalStyledDemo.js", "defaultCodeOpen": true}}
```

Author: It felt weird to me too that I am using the `experimentalStyled()` on this guide, but not sure what is the best place for it. Regarding the order, I agree. 

Not sure if the core is the best place for exporting the `experimentalStyled()` but it definitely depends on lots of core things, especially the theme structure that we use..

## Using standalone system utilities

> If you wouldn't like to use the `sx` prop on your custom components, but only some bits from the system, you can use and combine the different style functions available as part of the system.

Reviewer: ```suggestion
If you would prefer not to use the `sx` prop in your custom components, but only some parts of the system, you can use and combine the different style functions available.
```

Should we say why might a developer prefer this? (Better performance, smaller bundle size... ?)

Author: Done

> If you wouldn't like to use the `sx` prop on your custom components, but only some bits from the system, you can use and combine the different style function available as part of the system.

Reviewer 2: ```suggestion
If you wouldn't like to use the `sx` prop on your custom components, but only some bits from the system, you can use and combine the different style functions available as part of the system.
```
